opm/simulators/linalg/ISTLSolver.hpp

```diff
@@ -495,40 +495,45 @@ std::unique_ptr<Matrix> blockJacobiAdjacency(const Grid& grid,
             if (!flexibleSolver_[activeSolverNum_].solver_) {
                 return true;
             }
-            if (this->parameters_[activeSolverNum_].cpr_reuse_setup_ == 0) {
-                // Always recreate solver.
-                return true;
-            }
-            if (this->parameters_[activeSolverNum_].cpr_reuse_setup_ == 1) {
-                // Recreate solver on the first iteration of every timestep.
-                const int newton_iteration = this->simulator_.model().newtonMethod().numIterations();
-                return newton_iteration == 0;
-            }
-            if (this->parameters_[activeSolverNum_].cpr_reuse_setup_ == 2) {
-                // Recreate solver if the last solve used more than 10 iterations.
-                return this->iterations() > 10;
-            }
-            if (this->parameters_[activeSolverNum_].cpr_reuse_setup_ == 3) {
-                // Recreate solver if the last solve used more than 10 iterations.
-                return false;
-            }
-            if (this->parameters_[activeSolverNum_].cpr_reuse_setup_ == 4) {
-                // Recreate solver every 'step' solve calls.
-                const int step = this->parameters_[activeSolverNum_].cpr_reuse_interval_;
-                const bool create = ((solveCount_ % step) == 0);
-                return create;
-            }
-
-            // If here, we have an invalid parameter.
-            const bool on_io_rank = (simulator_.gridView().comm().rank() == 0);
-            std::string msg = "Invalid value: " + std::to_string(this->parameters_[activeSolverNum_].cpr_reuse_setup_)
-                + " for --cpr-reuse-setup parameter, run with --help to see allowed values.";
-            if (on_io_rank) {
-                OpmLog::error(msg);
+            // For CPR/AMG based preconditioners we have the option to recreate the linear solver
+            auto linsolver = this->parameters_[activeSolverNum_].linsolver_;
+            if (linsolver == "cpr" || linsolver == "cprw" || linsolver == "cpr"||
+                linsolver == "cpr_quasiimpes" || linsolver == "cpr_trueimpes" ||
+                linsolver == "cpr_trueimpesanalytic" || linsolver == "amg" ||
+                linsolver == "hybrid") {
```

Comment:
I think this condition is not precise enough. If you use an explicit .json file to set up your linear solver then this will not be checked, and you will never recreate the solver. I suggest instead making the opposite check: if using linsolver equal to "ilu0" or "dilu" then skip this, else do this. That would reproduce results for existing runs with .json files.

Corner case: if you then pass in a .json file specifying a preconditioner with "perfect update" such as ilu0 or dilu, and nothing else, then by default it would be needlessly recreated. However, then we have an expert user that can be expected to turn this off manually if required.

Better but more involved solution: extend the PreconditionerWithUpdate interface to provide a new method `bool hasPerfectUpdate()`, then skip this part if that returns true.

Reply:
I think the more involved solution extending the preconditionerWithUpdate would also make the code more consistent when maintained. Adding a new preconditioner would force the developer to find out whether or not this particular preconditioner must be reconstructed. Right now one could add a new preconditioner and be unaware that one must also update this particular if sentence in another file.

Reply:
See the proposed solution, which implements this.